In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import statsmodels.api as sm
print('20127250 \nHo Thi Thanh Ngan')

20127250 
Ho Thi Thanh Ngan


Câu hỏi 1: Trường ĐH KHTN đã thực hiện điều tra ngẫu nhiên một mẫu gồm các cựu sinh 
viên để hiểu về sự liên hệ giữa điểm số của sinh viên ra trường và mức lương khởi điểm.
Kiểm tra sự tương quan giữa hai biến số.

Dữ liệu: Tập dữ liệu bao gồm điểm GPA khi ra trường của sinh viên và mức lương khởi 
điểm tính bằng USD.

In [82]:
df = pd.read_csv("https://raw.githubusercontent.com/Nganho22/rvPython/main/Salary.csv", header=1, index_col=None)
df.head()

#df = pd.read_excel("film_sales.xlsx")
#df.to_csv('film_sales.csv', index=False)

Salary  GPA
0     300  8.5
1     400  8.0
2     400  6.7
3     890  9.3
4     295  7.5

In [83]:
DT = pd.DataFrame(df)
print("Tong so dong du lieu la: ", DT.shape[0])
column_types = DT.dtypes
print("Danh sach cac columns va type cua chung")
print(column_types)

Tong so dong du lieu la:  11
Danh sach cac columns va type cua chung
Salary      int64
GPA       float64
dtype: object


In [84]:
#Xu ly dulicate
DT.drop_duplicates()
print("Tong so dong du lieu sau khi xu ly duplicate la: ", DT.shape[0])

Tong so dong du lieu sau khi xu ly duplicate la:  11


In [85]:
#Xac dinh missing values
for column in DT.columns:
  per = DT[column].isnull().mean()
  print(f'{column}: {round(per*100,2)}%')

Salary: 0.0%
GPA: 0.0%


a) Cài đặt code đánh giá mức độ tương quan thông qua mô hình hồi quy. Cần giải thích 
rõ các thông số thu được từ kết quả

#1 có add_consitant: 

    salart = a + b*GPA

In [86]:
x = df['GPA']
x = sm.add_constant(x)
y = df['Salary']
model = sm.OLS(y,x)
results = model.fit()
print(results.summary())

                            OLS Regression Results                            
Dep. Variable:                 Salary   R-squared:                       0.193
Model:                            OLS   Adj. R-squared:                  0.103
Method:                 Least Squares   F-statistic:                     2.148
Date:                Sun, 17 Dec 2023   Prob (F-statistic):              0.177
Time:                        23:50:27   Log-Likelihood:                -73.071
No. Observations:                  11   AIC:                             150.1
Df Residuals:                       9   BIC:                             150.9
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const       -208.7084    454.356     -0.459      0.6

c:\Users\user\AppData\Local\Programs\Python\Python310\lib\site-packages\scipy\stats\_stats_py.py:1477: UserWarning: kurtosistest only valid for n>=20 ... continuing anyway, n=11
  warnings.warn("kurtosistest only valid for n>=20 ... continuing "



#test 1: salary = a + b*GPA

    R-squared: 0.193
    F-statistic: 2.148
    Prob: 0.17 hơi lớn thì với dataset không có ý nghĩa

                    p-value
    const(a)        0.657
    GPA  (b)        0.177

#2 bỏ add_consitant:
    salary = b*GPA

In [87]:
x = df['GPA']
y = df['Salary']
model = sm.OLS(y,x)
results = model.fit()
print(results.summary())

                                 OLS Regression Results                                
Dep. Variable:                 Salary   R-squared (uncentered):                   0.857
Model:                            OLS   Adj. R-squared (uncentered):              0.842
Method:                 Least Squares   F-statistic:                              59.77
Date:                Sun, 17 Dec 2023   Prob (F-statistic):                    1.59e-05
Time:                        23:50:27   Log-Likelihood:                         -73.198
No. Observations:                  11   AIC:                                      148.4
Df Residuals:                      10   BIC:                                      148.8
Df Model:                           1                                                  
Covariance Type:            nonrobust                                                  
                 coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------

c:\Users\user\AppData\Local\Programs\Python\Python310\lib\site-packages\scipy\stats\_stats_py.py:1477: UserWarning: kurtosistest only valid for n>=20 ... continuing anyway, n=11
  warnings.warn("kurtosistest only valid for n>=20 ... continuing "


#test 2: salary = b*GPA

    R-squared: 0.857
    F-statistic: 59.77
    Prob: 1.59E-05 tương đối bé và ổn hơn

                    P-value
    GPA (b)         0.000

    => chọn được mô hình : salary = 57.7115*GPA

In [88]:
print('Parameters: ', results.params)
print('R2: ', results.rsquared)

Parameters:  GPA    57.71152
dtype: float64
R2:  0.8566619532992307


b) Dự đoán mức lương khởi điểm cho sinh viên có điểm trung bình khi học đại học là 8.0

In [89]:
print('salary = 57.7115 * GPA')
print('GPA = 8.0')
print('Then predicted Salary: ', 57.7115*8.0)

salary = 57.7115 * GPA
GPA = 8.0
Then predicted Salary:  461.692


Câu hỏi 2: Sử dụng hồi quy tuyến tính để dự đoán Tổng sản phẩm quốc nội (GDP) của 
một quốc gia dựa trên các yếu tố như đầu tư công cộng, tiêu dùng cá nhân, và tỷ lệ thất 
nghiệp.

Dữ liệu: Tập dữ liệu bao gồm GDP hàng năm, đầu tư công cộng, tiêu dùng cá nhân, và 
tỷ lệ thất nghiệp cho một quốc gia trong vòng 10 năm qua

In [90]:
df1 = pd.read_csv("https://raw.githubusercontent.com/Nganho22/rvPython/main/GDP.csv", header=0, index_col=None)
df1.head()

Public Investment  Personal Spending  Unemployment Rate       GDP
0             829.18            5737.77               8.59  73228.77
1            1060.49            4136.05               4.72  47552.29
2             851.57            6061.11               8.94  66205.87
3            1080.47            4172.61               3.40  51984.68
4             957.45            4630.73               3.37  57237.32

In [91]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20 entries, 0 to 19
Data columns (total 4 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Public Investment  20 non-null     float64
 1   Personal Spending  20 non-null     float64
 2   Unemployment Rate  20 non-null     float64
 3   GDP                20 non-null     float64
dtypes: float64(4)
memory usage: 768.0 bytes


In [92]:
DT1 = pd.DataFrame(df1)
print("Tong so dong du lieu la: ", DT1.shape[0])
column_types = DT1.dtypes
print("Danh sach cac columns va type cua chung")
print(column_types)

Tong so dong du lieu la:  20
Danh sach cac columns va type cua chung
Public Investment    float64
Personal Spending    float64
Unemployment Rate    float64
GDP                  float64
dtype: object


In [93]:
#Xac dinh missing values
for column in DT1.columns:
  per = DT1[column].isnull().mean()
  print(f'{column}: {round(per*100,2)}%')

Public Investment: 0.0%
Personal Spending: 0.0%
Unemployment Rate: 0.0%
GDP: 0.0%


a) Xây dựng mô hình hồi quy tuyến tính sử dụng đầu tư công cộng, tiêu dùng cá nhân, và tỷ lệ thất nghiệp làm biến độc lập

#test 1: 

GDP = a + b*[Public Investment] + c*[Personal Spending] + d*[Unemployment Rate]

In [94]:
x = df1[['Public Investment','Personal Spending','Unemployment Rate']]
y = df1.GDP
x = sm.add_constant(x)
model = sm.OLS(y,x)
results = model.fit()
print(results.summary())

                            OLS Regression Results                            
Dep. Variable:                    GDP   R-squared:                       0.273
Model:                            OLS   Adj. R-squared:                  0.136
Method:                 Least Squares   F-statistic:                     2.001
Date:                Sun, 17 Dec 2023   Prob (F-statistic):              0.155
Time:                        23:50:28   Log-Likelihood:                -207.37
No. Observations:                  20   AIC:                             422.7
Df Residuals:                      16   BIC:                             426.7
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                        coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------
const              1.627e+04   2.01e+0

#test 1: GDP = a + b*[Public Investment] + c*[Personal Spending] + d*[Unemployment Rate]

    R-squared: 0.273
    F-statistic: 0.136
    Prob: 0.155  => tương đối lớn, có vẻ không phù hợp
                                P_value
const             (a)           0.429

Public Investment (b)           0.233

Personal Spending (c)           0.102

Unemployment Rate (d)           0.500

Có vẻ không phù hợp vì p_value lớn



#test 2: 

GDP = b*[Public Investment] + c*[Personal Spending] + d*[Unemployment Rate]

In [95]:
x = df1[['Public Investment','Personal Spending','Unemployment Rate']]
y = df1.GDP
model = sm.OLS(y,x)
results = model.fit()
print(results.summary())

                                 OLS Regression Results                                
Dep. Variable:                    GDP   R-squared (uncentered):                   0.984
Model:                            OLS   Adj. R-squared (uncentered):              0.982
Method:                 Least Squares   F-statistic:                              355.8
Date:                Sun, 17 Dec 2023   Prob (F-statistic):                    1.56e-15
Time:                        23:50:28   Log-Likelihood:                         -207.77
No. Observations:                  20   AIC:                                      421.5
Df Residuals:                      17   BIC:                                      424.5
Df Model:                           3                                                  
Covariance Type:            nonrobust                                                  
                        coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------

#test 2: GDP = b*[Public Investment] + c*[Personal Spending] + d*[Unemployment Rate]

    R-squared: 0.984
    F-statistic: 355.8
    Prob: 1.56e-15  => quá bé, có vẻ phù hợp
                                P_value

Public Investment (b)           0.009

Personal Spending (c)           0.026

Unemployment Rate (d)           0.294

Tuy nhiên P_value tại quá lớn ở giá trị d, cần tìm 1 công thức phù hợp hơn.



#test 3:

    GDP = a + b*[Public Investment] + c*[Personal Spending]

In [96]:
x = df1[['Public Investment','Personal Spending']]
y = df1.GDP
x = sm.add_constant(x)
model = sm.OLS(y,x)
results = model.fit()
print(results.summary())

                            OLS Regression Results                            
Dep. Variable:                    GDP   R-squared:                       0.251
Model:                            OLS   Adj. R-squared:                  0.163
Method:                 Least Squares   F-statistic:                     2.851
Date:                Sun, 17 Dec 2023   Prob (F-statistic):             0.0855
Time:                        23:50:28   Log-Likelihood:                -207.66
No. Observations:                  20   AIC:                             421.3
Df Residuals:                      17   BIC:                             424.3
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                        coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------
const              2.145e+04   1.83e+0

#test 3: GDP = a + b*[Public Investment] + c*[Personal Spending]

    R-squared: 0.251
    F-statistic: 2.851
    Prob: 0.0855  => tương đối lớn, có vẻ không phù hợp
                                P_value

const             (a)           0.257

Public Investment (b)           0.304

Personal Spending (c)           0.040



#test 4: 

    GDP = b*[Public Investment] + c*[Personal Spending]

In [97]:
x = df1[['Public Investment','Personal Spending']]
y = df1.GDP
model = sm.OLS(y,x)
results = model.fit()

print(results.summary())


                                 OLS Regression Results                                
Dep. Variable:                    GDP   R-squared (uncentered):                   0.983
Model:                            OLS   Adj. R-squared (uncentered):              0.981
Method:                 Least Squares   F-statistic:                              528.0
Date:                Sun, 17 Dec 2023   Prob (F-statistic):                    1.04e-16
Time:                        23:50:29   Log-Likelihood:                         -208.44
No. Observations:                  20   AIC:                                      420.9
Df Residuals:                      18   BIC:                                      422.9
Df Model:                           2                                                  
Covariance Type:            nonrobust                                                  
                        coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------

#test 4: GDP = b*[Public Investment] + c*[Personal Spending]

    R-squared: 0.983
    F-statistic: 528.0
    Prob:1.04e-16  => tương đối nhỏ, có vẻ phù hợp và đồng thời R-squared cũng rất cao

                                P_value

Public Investment (b)           0.013

Personal Spending (c)           0.000

P_value cũng hợp lý, vì thế tôi lựa chọn đây là mô hình hợp lý.


GDP = 22.9247 * [Public Investment] + 7.2631 * [Personal Spending]

GDP phụ thuộc vào [Public Investment] & [Persional Spending], nhưng không phụ thuộc vào [Unemployment Rate]

Để dự đoán GDP trong năm tiếp theo dựa trên các ước lượng từ mô hình, vì không có giá trị dữ liệu từ năm tiếp theo tôi sẽ tạm thời lấy giá trị mean từ tập dữ liệu.

Xử dụng LinearRegression để dự đoán

In [98]:
from sklearn.linear_model import LinearRegression

x = df1[['Public Investment','Personal Spending']]
y = df1.GDP
model = LinearRegression()
model.fit(x,y)
ave_investment = df1['Public Investment'].mean()
ave_Spending = df1['Personal Spending'].mean()

new_data = pd.DataFrame({'Public Investment': [ave_investment], 'Personal Spending': [ave_Spending]})
predicted_gdp = model.predict(new_data)
print("Dự đoán GDP của năm sau thông qua LinearRegression là: ", predicted_gdp[0])

Dự đoán GDP của năm sau thông qua LinearRegression là:  62125.1915


Xử dụng mô hình vừa tìm được ở trên

GDP = 22.9247 * ave_investment + 7.2631 * ave_Spending

In [99]:
GDP = 22.9247 * ave_investment + 7.2631 * ave_Spending
print("Dự đoán GDP của năm sau thông qua mô hình vừa tìm được là : ",22.9247 * ave_investment + 7.2631 * ave_Spending )

Dự đoán GDP của năm sau thông qua mô hình vừa tìm được là :  61895.077714600004


Xử dụng phép dự đoán của OLS

In [100]:
x = df1[['Public Investment','Personal Spending']]
y = df1.GDP
model = sm.OLS(y,x)
results = model.fit()
new_data = pd.DataFrame({'Public Investment': [ave_investment], 'Personal Spending': [ave_Spending]})
predicted_gdp = results.predict(new_data)
print('Dự đoán GDP của năm sau thông qua phép dự đoán của OLS là :',predicted_gdp.values[0])


Dự đoán GDP của năm sau thông qua phép dự đoán của OLS là : 61894.95549215603


Từ 3 phép dự đoán trên, ta có thể thấy GDP có vẻ sẽ dao động từ 61894 -> 62125

Câu hỏi 3: 

Sử dụng lại dataset ở BTLT 2 (E2.3), dùng phương pháp hồi quy điền vào missing value cho cột overall scores

In [101]:
df2 = pd.read_csv("https://raw.githubusercontent.com/Nganho22/20127250_BT2/master/World%20University%20Rankings%202023.csv")
df2.head()


University Rank                     Name of University        Location  \
0               1                   University of Oxford  United Kingdom   
1               2                     Harvard University   United States   
2               3                University of Cambridge  United Kingdom   
3               3                    Stanford University   United States   
4               5  Massachusetts Institute of Technology   United States   

  No of student  No of student per staff International Student  \
0        20,965                     10.6                   42%   
1        21,887                      9.6                   25%   
2        20,185                     11.3                   39%   
3        16,164                      7.1                   24%   
4        11,415                      8.2                   33%   

  Female:Male Ratio OverAll Score  Teaching Score  Research Score  \
0           48 : 52          96.4            92.3            99.7   
1           50 : 50          95.2            94.8            99.0   
2           47 : 53          94.8            90.9            99.5   
3           46 : 54          94.8            94.2            96.7   
4           40 : 60          94.2            90.7            93.6   

   Citations Score  Industry Income Score  International Outlook Score  
0             99.0                   74.9                         96.2  
1             99.3                   49.5                         80.5  
2             97.0                   54.2                         95.8  
3             99.8                   65.0                         79.8  
4             99.8                   90.9                         89.3

Trước hết tôi sẽ xử lý missing ở các cột khác OverAll Score

In [102]:
#df = pd.DataFrame(df2)
df2 = df2.drop_duplicates(subset='Name of University', keep='first')
df2 = df2[['OverAll Score','Teaching Score','Research Score','Citations Score', 'Industry Income Score','International Outlook Score']]

In [103]:
df2.head()


OverAll Score  Teaching Score  Research Score  Citations Score  \
0          96.4            92.3            99.7             99.0   
1          95.2            94.8            99.0             99.3   
2          94.8            90.9            99.5             97.0   
3          94.8            94.2            96.7             99.8   
4          94.2            90.7            93.6             99.8   

   Industry Income Score  International Outlook Score  
0                   74.9                         96.2  
1                   49.5                         80.5  
2                   54.2                         95.8  
3                   65.0                         79.8  
4                   90.9                         89.3

In [104]:
print('Xem lại tỷ lệ missing ở mỗi cột')
for column in df2.columns:
  per = df2[column].isnull().mean()
  print(f'{column}: {round(per*100,2)}%')

Tong_missing_values = df2.isna().sum()
print("\nTong gia tri bi missing o moi column la:")
print(Tong_missing_values)

Xem lại tỷ lệ missing ở mỗi cột
OverAll Score: 22.92%
Teaching Score: 22.92%
Research Score: 22.92%
Citations Score: 22.92%
Industry Income Score: 22.92%
International Outlook Score: 22.92%

Tong gia tri bi missing o moi column la:
OverAll Score                  512
Teaching Score                 512
Research Score                 512
Citations Score                512
Industry Income Score          512
International Outlook Score    512
dtype: int64


In [105]:
print("Kiểm tra xem các cột còn lại có cùng missing với nhau")
check = df2[df2['OverAll Score'].isnull()]
count = check.isnull().sum()
print(count)

Kiểm tra xem các cột còn lại có cùng missing với nhau
OverAll Score                  512
Teaching Score                 512
Research Score                 512
Citations Score                512
Industry Income Score          512
International Outlook Score    512
dtype: int64


In [106]:
print('Drop toàn bộ cột chứa null')
df2.dropna(subset='OverAll Score', inplace=True)
Tong_missing_values = df2.isna().sum()
print("Tong gia tri bi missing o moi column la:")
print(Tong_missing_values)

Drop toàn bộ cột chứa null
Tong gia tri bi missing o moi column la:
OverAll Score                  0
Teaching Score                 0
Research Score                 0
Citations Score                0
Industry Income Score          0
International Outlook Score    0
dtype: int64


In [107]:
column_types = df2.dtypes
print("Danh sach cac columns va type cua chung")
print(column_types)


Danh sach cac columns va type cua chung
OverAll Score                   object
Teaching Score                 float64
Research Score                 float64
Citations Score                float64
Industry Income Score          float64
International Outlook Score    float64
dtype: object


Tôi sẽ chuyển đổi cột OverAll Score thành numeric, nếu tồn tại ký từ không phải là từ '0' -> '9' thì sẽ null

In [108]:
df2['OverAll Score'] = pd.to_numeric(df2['OverAll Score'], errors='coerce')


In [109]:
Tong_missing_values = df2.isna().sum()
print("Tong gia tri bi missing o moi column la:")
print(Tong_missing_values)

Tong gia tri bi missing o moi column la:
OverAll Score                  1522
Teaching Score                    0
Research Score                    0
Citations Score                   0
Industry Income Score             0
International Outlook Score       0
dtype: int64


Tôi sẽ tiến hành lấy những cột không bị null ra để tìm mô hình thích hợp

In [110]:
df3 = df2[df2['OverAll Score'].notnull()]
df3.head()

OverAll Score  Teaching Score  Research Score  Citations Score  \
0           96.4            92.3            99.7             99.0   
1           95.2            94.8            99.0             99.3   
2           94.8            90.9            99.5             97.0   
3           94.8            94.2            96.7             99.8   
4           94.2            90.7            93.6             99.8   

   Industry Income Score  International Outlook Score  
0                   74.9                         96.2  
1                   49.5                         80.5  
2                   54.2                         95.8  
3                   65.0                         79.8  
4                   90.9                         89.3

#test 1: 

[OverAll Score] = a + b*[Teaching Score] + c*[Research Score] + d*[Citations Score] + e*[Industry Income Score] + f*[International Outlook Score]

In [111]:
x = df3[['Teaching Score','Research Score','Citations Score','Industry Income Score','International Outlook Score']]
y = df3['OverAll Score']
x = sm.add_constant(x)
model = sm.OLS(y,x)
results = model.fit()
print(results.summary())

                            OLS Regression Results                            
Dep. Variable:          OverAll Score   R-squared:                       1.000
Model:                            OLS   Adj. R-squared:                  1.000
Method:                 Least Squares   F-statistic:                 4.557e+06
Date:                Sun, 17 Dec 2023   Prob (F-statistic):               0.00
Time:                        23:50:31   Log-Likelihood:                 404.43
No. Observations:                 200   AIC:                            -796.9
Df Residuals:                     194   BIC:                            -777.1
Df Model:                           5                                         
Covariance Type:            nonrobust                                         
                                  coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------------------
const             

có vẻ tương đối phù hợp vì Prob = 0
R-squared:                       1.000
F-statistic:                 4.557e+06
Prob (F-statistic):               0.00

tuy nhiên p_value của const lại quá lớn 0.694, nên t sẽ tiến hành lựa chọn mô hình khác

#test 2: 

[OverAll Score] = b*[Teaching Score] + c*[Research Score] + d*[Citations Score] + e*[Industry Income Score] + f*[International Outlook Score]

In [112]:
x = df3[['Teaching Score','Research Score','Citations Score','Industry Income Score','International Outlook Score']]
y = df3['OverAll Score']
model = sm.OLS(y,x)
results = model.fit()
print(results.summary())

                                 OLS Regression Results                                
Dep. Variable:          OverAll Score   R-squared (uncentered):                   1.000
Model:                            OLS   Adj. R-squared (uncentered):              1.000
Method:                 Least Squares   F-statistic:                          1.728e+08
Date:                Sun, 17 Dec 2023   Prob (F-statistic):                        0.00
Time:                        23:50:31   Log-Likelihood:                          404.35
No. Observations:                 200   AIC:                                     -798.7
Df Residuals:                     195   BIC:                                     -782.2
Df Model:                           5                                                  
Covariance Type:            nonrobust                                                  
                                  coef    std err          t      P>|t|      [0.025      0.975]
------------------------

Hoàn toàn phù hợp vì R-squared = 1, Prob = 0.00


                                        p_value

Teaching Score              (b)         0.000

Research Score              (c)         0.000

Citations Score             (d)         0.000

Industry Income Score       (e)         0.000

International Outlook Score (f)         0.000


Vậy mô hình thích hợp nhất là: 

[OverAll Score] = 0.2998*[Teaching Score] + 0.3003*[Research Score] + 0.3001*[Citations Score] + 0.025*[Industry Income Score] + 0.0749*[International Outlook Score]

Điền lại các giá trị missing

In [113]:
coef ={
    'Teaching Score': 0.2998,
    'Research Score': 0.3003,
    'Citations Score': 0.3001,
    'Industry Income Score': 0.025,
    'International Outlook Score': 0.0749
}
calculate = (
    coef['Teaching Score'] * df2['Teaching Score'] +
    coef['Research Score'] * df2['Research Score'] +
    coef['Citations Score'] * df2['Citations Score'] +
    coef['Industry Income Score'] * df2['Industry Income Score'] +
    coef['International Outlook Score'] * df2['International Outlook Score']
)

df2['OverAll Score'].fillna(calculate, inplace=True)

print('Xem lại tỷ lệ missing ở mỗi cột sau khi xử lý missing là:')
for column in df2.columns:
  per = df2[column].isnull().mean()
  print(f'{column}: {round(per*100,2)}%')

Xem lại tỷ lệ missing ở mỗi cột sau khi xử lý missing là:
OverAll Score: 0.0%
Teaching Score: 0.0%
Research Score: 0.0%
Citations Score: 0.0%
Industry Income Score: 0.0%
International Outlook Score: 0.0%
